Improve documentation and speedup isempty

bkamins opened the pull request:
@nalimilan :

1. Your recent comment prompted me to add `isempty` documentation as in DataFrames.jl when we say that data frame is empty it means that it has 0 rows (but can have some columns). I have also simplified `isempty` implementation as currently data frame that has positive number of rows must have some columns.
2. I have also split duplicate definitions in documentation that we discussed we should do.

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -357,7 +357,12 @@ function Base.size(df::AbstractDataFrame, i::Integer)
 
-Base.isempty(df::AbstractDataFrame) = size(df, 1) == 0 || size(df, 2) == 0
+"""
+    isempty(::AbstractDataFrame)
```
```suggestion
    isempty(df::AbstractDataFrame)
```

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -357,7 +357,12 @@ function Base.size(df::AbstractDataFrame, i::Integer)
+"""
+    isempty(::AbstractDataFrame)
+
+Return `true` if data frame has zero rows.
```
```suggestion
Return `true` if data frame `df` has zero rows, and `false` otherwise.
```

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -1373,22 +1378,17 @@ end
+See also: [`unique!`](@ref) [`nonunique`](@ref).
 
 # Arguments
 - `df` : the AbstractDataFrame
```
While you're at it, on the line below backticks are missing. Same below.

bkamins replied:
fixed

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
````diff
@@ -1438,6 +1438,52 @@ julia> unique(df, 2)
    2 │     2      2
+```
+"""
+unique
````
Attach this to the actual method now that it's possible?

bkamins replied:
done

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
````diff
@@ -2120,34 +2166,43 @@ Base.parentindices(adf::AbstractDataFrame) = axes(adf)
+```
+
+"""
+function nrow end
````
Same here.

bkamins replied:
done

bkamins commented:
When applying your suggestions I cleaned up the code a bit (in particular merged `ncol` definition and removed one `unique!` method that was not needed).

bkamins commented:
It turns out that `Base.unique!(df::AbstractDataFrame, cols::AbstractVector)` definition is required (resolves dispatch ambiguity)

bkamins commented:
@nalimilan - this PR should be clean now.

nalimilan commented on src/dataframe/dataframe.jl:
````diff
@@ -427,9 +427,23 @@ _onebased_check_error(i, col) =
+```
+""" # note: these type assertions are required to pass tests
 nrow(df::DataFrame) = ncol(df) > 0 ? length(_columns(df)[1])::Int : 0
-ncol(df::DataFrame) = length(index(df))
````
Why do you separate `nrow` from `ncol` definitions?

bkamins replied:
Because `ncol` definition is generic and works for any `AbstractDataFrame` so it goes into abstractdataframe/abstractdataframe.jl.
While `nrow` needs a separate definition for `DataFrame` and `SubDataFrame` (and they are respectively in dataframe/dataframe.jl and subdataframe/subdataframe.jl).

This separation is a consequence of your request to attach docsting of `nrow` to actual method definition (and not to `function nrow end` placeholder that was in abstractdataframe/abstractdataframe.jl in my initial version of this PR).

bkamins replied:
(I preferred having `function nrow end` and a docstring in abstractdataframe/abstractdataframe.jl so if you now agree to do this this way I will change the PR)

bkamins commented:
Thank you!
